File: src/csharp/Intel/Iced.UnitTests/Intel/FormatterTests/FormatterTestUtils.cs
/*
Copyright (C) 2018-2019 de4dot@gmail.com

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

#if GAS || INTEL || MASM || NASM
using System;
using Iced.Intel;
using Xunit;

namespace Iced.UnitTests.Intel.FormatterTests {
	static class FormatterTestUtils {
		public static void FormatTest(int bitness, string hexBytes, Code code, DecoderOptions options, string formattedString, Formatter formatter) {
			var decoder = CreateDecoder(bitness, hexBytes, options, out ulong nextRip);
			var instruction = decoder.Decode();
			Assert.Equal(code, instruction.Code);
			Assert.Equal((ushort)nextRip, instruction.IP16);
			Assert.Equal((uint)nextRip, instruction.IP32);
			Assert.Equal(nextRip, instruction.IP);
			nextRip += (uint)instruction.Length;
			Assert.Equal(nextRip, decoder.IP);
			Assert.Equal((ushort)nextRip, instruction.NextIP16);
			Assert.Equal((uint)nextRip, instruction.NextIP32);
			Assert.Equal(nextRip, instruction.NextIP);
			FormatTest(instruction, formattedString, formatter);
		}

		public static void FormatTest(in Instruction instruction, string formattedString, Formatter formatter) {
			var output = new StringOutput();

			formatter.Format(instruction, output);
			var actualFormattedString = output.ToStringAndReset();
#pragma warning disable xUnit2006 // Do not use invalid string equality check
			// Show the full string without ellipses by using Equal<string>() instead of Equal()
			Assert.Equal<string>(formattedString, actualFormattedString);
#pragma warning restore xUnit2006 // Do not use invalid string equality check

			formatter.FormatMnemonic(instruction, output);
			var mnemonic = output.ToStringAndReset();
			int opCount = formatter.GetOperandCount(instruction);
			var operands = opCount == 0 ? Array.Empty<string>() : new string[opCount];
			for (int i = 0; i < operands.Length; i++) {
				formatter.FormatOperand(instruction, output, i);
				operands[i] = output.ToStringAndReset();
			}
			output.Write(mnemonic, FormatterTextKind.Text);
			if (operands.Length > 0) {
				output.Write(" ", FormatterTextKind.Text);
				for (int i = 0; i < operands.Length; i++) {
					if (i > 0)
						formatter.FormatOperandSeparator(instruction, output);
					output.Write(operands[i], FormatterTextKind.Text);
				}
			}
			actualFormattedString = output.ToStringAndReset();
			Assert.Equal(formattedString, actualFormattedString);

			formatter.FormatAllOperands(instruction, output);
			var allOperands = output.ToStringAndReset();
			actualFormattedString = allOperands.Length == 0 ? mnemonic : mnemonic + " " + allOperands;
			Assert.Equal(formattedString, actualFormattedString);
		}

		public static void SimpleFormatTest(int bitness, string hexBytes, Code code, DecoderOptions options, string formattedString, Formatter formatter, Action<Decoder> initDecoder) {
			var decoder = CreateDecoder(bitness, hexBytes, options, out _);
			initDecoder?.Invoke(decoder);
			var nextRip = decoder.IP;
			var instruction = decoder.Decode();
			Assert.Equal(code, instruction.Code);
			Assert.Equal((ushort)nextRip, instruction.IP16);
			Assert.Equal((uint)nextRip, instruction.IP32);
			Assert.Equal(nextRip, instruction.IP);
			nextRip += (uint)instruction.Length;
			Assert.Equal(nextRip, decoder.IP);
			Assert.Equal((ushort)nextRip, instruction.NextIP16);
			Assert.Equal((uint)nextRip, instruction.NextIP32);
			Assert.Equal(nextRip, instruction.NextIP);

			var output = new StringOutput();
			formatter.Format(instruction, output);
			var actualFormattedString = output.ToStringAndReset();
#pragma warning disable xUnit2006 // Do not use invalid string equality check
			// Show the full string without ellipses by using Equal<string>() instead of Equal()
			Assert.Equal<string>(formattedString, actualFormattedString);
#pragma warning restore xUnit2006 // Do not use invalid string equality check
		}

		static Decoder CreateDecoder(int bitness, string hexBytes, DecoderOptions options, out ulong rip) {
			var codeReader = new ByteArrayCodeReader(hexBytes);
			var decoder = Decoder.Create(bitness, codeReader, options);
			rip = bitness switch {
				16 => DecoderConstants.DEFAULT_IP16,
				32 => DecoderConstants.DEFAULT_IP32,
				64 => DecoderConstants.DEFAULT_IP64,
				_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
			};
			Assert.Equal(bitness, decoder.Bitness);
			decoder.IP = rip;
			return decoder;
		}

		public static void TestFormatterDoesNotThrow(Formatter formatter) {
			var output = new StringOutput();
			foreach (var info in DecoderTests.DecoderTestUtils.GetDecoderTests(includeOtherTests: true, includeInvalid: true)) {
				var decoder = CreateDecoder(info.Bitness, info.HexBytes, info.Options, out _);
				decoder.Decode(out var instruction);
				formatter.Format(instruction, output);
				output.Reset();
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/fmt/FormatterTestUtils.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.fmt;

import java.util.function.Consumer;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.*;

import com.github.icedland.iced.x86.HexUtils;
import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.dec.ByteArrayCodeReader;
import com.github.icedland.iced.x86.dec.Decoder;
import com.github.icedland.iced.x86.dec.DecoderTestInfo;
import com.github.icedland.iced.x86.dec.DecoderTestUtils;
import com.github.icedland.iced.x86.fmt.fast.FastFormatter;
import com.github.icedland.iced.x86.fmt.fast.FastStringOutput;

final class FormatterTestUtils {
	static void simpleFormatTest(int bitness, String hexBytes, long ip, int code, int options, String formattedString,
			Function<Instruction, String> format, Consumer<Decoder> initDecoder) {
		Decoder decoder = createDecoder(bitness, hexBytes, ip, options);
		if (initDecoder != null)
			initDecoder.accept(decoder);
		long nextRip = decoder.getIP();
		Instruction instruction = decoder.decode();
		assertEquals(code, instruction.getCode());
		assertEquals((short)nextRip, instruction.getIP16());
		assertEquals((int)nextRip, instruction.getIP32());
		assertEquals(nextRip, instruction.getIP());
		nextRip += instruction.getLength();
		assertEquals(nextRip, decoder.getIP());
		assertEquals((short)nextRip, instruction.getNextIP16());
		assertEquals((int)nextRip, instruction.getNextIP32());
		assertEquals(nextRip, instruction.getNextIP());

		String actualFormattedString = format.apply(instruction);
		assertEquals(formattedString, actualFormattedString);
	}

	private static Decoder createDecoder(int bitness, String hexBytes, long ip, int options) {
		ByteArrayCodeReader codeReader = new ByteArrayCodeReader(HexUtils.toByteArray(hexBytes));
		Decoder decoder = new Decoder(bitness, codeReader, options);
		assertEquals(bitness, decoder.getBitness());
		decoder.setIP(ip);
		return decoder;
	}

	static void formatTest(int bitness, String hexBytes, long ip, int code, int options, String formattedString, Formatter formatter) {
		Decoder decoder = createDecoder(bitness, hexBytes, ip, options);
		long nextRip = decoder.getIP();
		Instruction instruction = decoder.decode();
		assertEquals(code, instruction.getCode());
		assertEquals((short)nextRip, instruction.getIP16());
		assertEquals((int)nextRip, instruction.getIP32());
		assertEquals(nextRip, instruction.getIP());
		nextRip += instruction.getLength();
		assertEquals(nextRip, decoder.getIP());
		assertEquals((short)nextRip, instruction.getNextIP16());
		assertEquals((int)nextRip, instruction.getNextIP32());
		assertEquals(nextRip, instruction.getNextIP());
		formatTest(instruction, formattedString, formatter);
	}

	static void formatTest(Instruction instruction, String formattedString, Formatter formatter) {
		StringOutput output = new StringOutput();

		formatter.format(instruction, output);
		String actualFormattedString = output.toStringAndReset();
		assertEquals(formattedString, actualFormattedString);

		formatter.formatMnemonic(instruction, output);
		String mnemonic = output.toStringAndReset();
		int opCount = formatter.getOperandCount(instruction);
		String[] operands = new String[opCount];
		for (int i = 0; i < operands.length; i++) {
			formatter.formatOperand(instruction, output, i);
			operands[i] = output.toStringAndReset();
		}
		output.write(mnemonic, FormatterTextKind.TEXT);
		if (operands.length > 0) {
			output.write(" ", FormatterTextKind.TEXT);
			for (int i = 0; i < operands.length; i++) {
				if (i > 0)
					formatter.formatOperandSeparator(instruction, output);
				output.write(operands[i], FormatterTextKind.TEXT);
			}
		}
		actualFormattedString = output.toStringAndReset();
		assertEquals(formattedString, actualFormattedString);

		formatter.formatAllOperands(instruction, output);
		String allOperands = output.toStringAndReset();
		actualFormattedString = allOperands.length() == 0 ? mnemonic : mnemonic + " " + allOperands;
		assertEquals(formattedString, actualFormattedString);
	}

	static void simpleFormatTest(int bitness, String hexBytes, long ip, int code, int options, String formattedString, Formatter formatter,
			Consumer<Decoder> initDecoder) {
		Function<Instruction, String> format = (Instruction instruction) -> {
			StringOutput output = new StringOutput();
			formatter.format(instruction, output);
			return output.toStringAndReset();
		};
		simpleFormatTest(bitness, hexBytes, ip, code, options, formattedString, format, initDecoder);
	}

	static void testFormatterDoesNotThrow(Formatter formatter) {
		StringOutput output = new StringOutput();
		for (DecoderTestInfo info : DecoderTestUtils.getDecoderTests(true, true)) {
			Decoder decoder = createDecoder(info.bitness, info.hexBytes, info.ip, info.options);
			Instruction instruction = new Instruction();
			decoder.decode(instruction);
			formatter.format(instruction, output);
			output.reset();
		}
	}

	static void formatTest(int bitness, String hexBytes, long ip, int code, int options, String formattedString, FastFormatter formatter) {
		Decoder decoder = createDecoder(bitness, hexBytes, ip, options);
		long nextRip = decoder.getIP();
		Instruction instruction = decoder.decode();
		assertEquals(code, instruction.getCode());
		assertEquals((short)nextRip, instruction.getIP16());
		assertEquals((int)nextRip, instruction.getIP32());
		assertEquals(nextRip, instruction.getIP());
		nextRip += instruction.getLength();
		assertEquals(nextRip, decoder.getIP());
		assertEquals((short)nextRip, instruction.getNextIP16());
		assertEquals((int)nextRip, instruction.getNextIP32());
		assertEquals(nextRip, instruction.getNextIP());
		formatTest(instruction, formattedString, formatter);
	}

	static void formatTest(Instruction instruction, String formattedString, FastFormatter formatter) {
		FastStringOutput output = new FastStringOutput();

		formatter.format(instruction, output);
		String actualFormattedString = output.toString();
		assertEquals(formattedString, actualFormattedString);
	}

	static void simpleFormatTest(int bitness, String hexBytes, long ip, int code, int options, String formattedString, FastFormatter formatter,
			Consumer<Decoder> initDecoder) {
		Function<Instruction, String> format = (Instruction instruction) -> {
			FastStringOutput output = new FastStringOutput();
			formatter.format(instruction, output);
			return output.toString();
		};
		simpleFormatTest(bitness, hexBytes, ip, code, options, formattedString, format, initDecoder);
	}
}
